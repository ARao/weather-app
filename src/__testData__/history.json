{
    "location": {},
    "forecast": {
        "forecastday": [
            {
                "date": "2019-01-04",
                "date_epoch": 1546560000,
                "day": {},
                "astro": {},
                "hour": []
            }
        ]
    }
}