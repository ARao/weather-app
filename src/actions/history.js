import { CITY_NAME, KEY, HISTORY_DAYS } from '../constants/appConst';
import { HISTORY_WEATHER } from '../constants/actionConst'
import { historyWeatherUrl } from '../constants/apiUrl'
import interceptor from '../interceptor'


const historyWeatherUrlParam = new URL(historyWeatherUrl.href);

historyWeatherUrlParam.searchParams.append('key', 'q', 'dt')

export const fetchHistories = () => dispatch => {
  historyWeatherUrlParam.searchParams.set('key', KEY)
  historyWeatherUrlParam.searchParams.set('q', CITY_NAME)

  let arr = [];
  for (let day of getNextDate()) {
    historyWeatherUrlParam.searchParams.set('dt', day)
    arr.push(fetch(historyWeatherUrlParam.href))
  }

  return Promise.all(arr)
    .then(histories => Promise.all(histories.map(history => history.json())))
    .then(histories => {
      dispatch({
        type: HISTORY_WEATHER,
        payload: histories,
        interceptor
      });
    })
    .catch(err => {
      console.log("error at historyAction ------->")
      console.log(err)
    })

}

function* getNextDate() {
  const today = new Date();
  today.setDate(today.getDate() - HISTORY_DAYS);
  for (let i = 0; i < HISTORY_DAYS; i++) {
    yield `${today.getDate()}-${today.getMonth() + 1}-${today.getFullYear()}`;
    today.setDate(today.getDate() + 1);
  }
  return;
}
