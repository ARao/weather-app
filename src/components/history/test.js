describe('History' , () => {
    it(' should be true ', ()=>{
        expect(true).toBeTruthy()
    })
})

