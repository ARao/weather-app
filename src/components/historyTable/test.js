import React from 'react';
import HistoryTable from './index'
import { shallow } from 'enzyme'
import history from '../../__testData__/history.json'


describe('HistoryTable', () => {
  const histories = [history]

  it('++++ render History table', () => {
    let wrapper = shallow(<HistoryTable histories={histories} />);
    expect(wrapper.length).toEqual(1);
  });

  it('++++ snapshot match', () => {
    let wrapper = shallow(<HistoryTable histories={histories} />);
    expect(wrapper).toMatchSnapshot();
  });
});
