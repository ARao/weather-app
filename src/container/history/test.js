import React from 'react';
import { shallow } from 'enzyme';
import ConnectedHistory from './index';
import { History } from './index'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import toJson from 'enzyme-to-json'
import history from '../../__testData__/history.json'

describe('History connectd component', () => {
  const fetchHistories = jest.fn(()=>Promise.resolve())

  const loaderShow = jest.fn();
  const loaderHide = jest.fn();

  const props = {
    fetchHistories,
    loaderShow,
    loaderHide
  }

  const initialState = {
    weather: {
      histories: [history]
    }
  }

  const middlewares = [ thunk ]
  const mockStore = configureMockStore(middlewares);
  let wrapper, store;

  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = shallow(<ConnectedHistory store={store} {...props} />)
  })

  it('++++ render the connected(SMART) component', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('++++ store should match', () => {
    expect(wrapper.props().histories).toEqual(initialState.weather.histories)
  });

  it('+++ snapshot match', () => {
    const snap = toJson(wrapper.dive());
    expect(snap).toMatchSnapshot();
  });
  it('++++ fetchHistories called', () =>{
    wrapper = shallow(<History { ...props } />)
    const instance = wrapper.instance();
    const spy = jest.spyOn(instance.props, 'fetchHistories');
    expect(spy).toHaveBeenCalled();
  })
  
  
});
