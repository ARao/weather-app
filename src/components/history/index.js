import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import BasicInfo from '../basicInfo'
import { fetchHistories } from '../../actions/historyActions'
import { Link } from 'react-router-dom'

class History extends Component {

  componentWillMount(){
    this.props.fetchHistories()
  }

  render() {
    const histories = this.props.histories && this.props.histories.map((history, index) =>{
          return (
            <tr  key={`${history.forecast.forecastday[0].date}--${index}`} >
              <td>{ history.forecast.forecastday[0].date }</td>
              <td>{history.forecast.forecastday[0].day.mintemp_c}</td>
              <td>{history.forecast.forecastday[0].day.mintemp_c}</td>
            </tr>
          )
    })
    return (
      <React.Fragment>
        <Link  to="/"><button className="btn btn-info mt-4" >Back</button></Link>
        <BasicInfo weather={this.props.histories && this.props.histories[0]}/>
        <hr/>
        <table className="table table-striped">
          <thead>
            <tr>
              <th scope="col" >Date</th>
              <th scope="col" >Min</th>
              <th scope="col" >Max</th>
            </tr>
          </thead>
          <tbody>
            {histories}
          </tbody>
        </table>
      </React.Fragment>
    );
  }
}

History.propTypes = {
  fetchHistories: PropTypes.func.isRequired,
  histories : PropTypes.array,
};

const mapStateToProps = state =>({
      histories : state.weather.histories,
});

export default connect(mapStateToProps, { fetchHistories })(History);
