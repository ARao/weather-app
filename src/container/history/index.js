import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import BasicInfo from '../../components/basicInfo';
import HistoryTable from '../../components/historyTable'
import { fetchHistories } from '../../actions/historyActions'
import { Link } from 'react-router-dom'

export class History extends Component {

  componentWillMount() {
    this.props.fetchHistories()
  }

  render() {
    return (
      <React.Fragment>
        <Link to="/"><button className="btn btn-info mt-4" >Back</button></Link>
        <BasicInfo weather={this.props.histories && this.props.histories[0]} />
        <hr />
        <HistoryTable histories={this.props.histories} />
      </React.Fragment>
    );
  }
}

History.propTypes = {
  fetchHistories: PropTypes.func.isRequired,
  histories: PropTypes.array,
};

const mapStateToProps = state => ({
  histories: state.weather.histories,
});

export default connect(mapStateToProps, { fetchHistories })(History);
