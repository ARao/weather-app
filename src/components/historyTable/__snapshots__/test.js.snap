// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryTable ++++ snapshot match 1`] = `
<table
  className="table table-striped"
>
  <thead>
    <tr>
      <th
        scope="col"
      >
        Date
      </th>
      <th
        scope="col"
      >
        Min
      </th>
      <th
        scope="col"
      >
        Max
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="2019-01-04--0"
    >
      <td>
        2019-01-04
      </td>
      <td />
      <td />
    </tr>
  </tbody>
</table>
`;
